provider/nbd_image_export.py

```diff
@@ -0,0 +1,170 @@
+                                                           'disk')
+        # Don't hotplug device driver
+        devices.pop()
+        self._node_name = devices[-1].get_qid()
+        self._image_devices = devices
+
+        logging.info("Plug devices(without image device driver)")
+        for dev in devices:
+            ret = self._vm.devices.simple_hotplug(dev, self._vm.monitor)
+            if not ret[1]:
+                raise exceptions.TestFail("Failed to hotplug device '%s': %s."
+                                          % (dev, ret[0]))
+
+    def hotplug_tls(self):
+        if self._image_params.get('nbd_unix_socket'):
+            logging.info('TLS is only supported with IP')
+        elif self._image_params.get('nbd_server_tls_creds'):
+            logging.info("Plug server tls creds device")
+            self._tls_creds_id = '%s_server_tls_creds' % self._tag
+            dev = qemu_devices.qdevices.QObject('tls-creds-x509')
+            dev.set_param("id", self._tls_creds_id)
+            dev.set_param("endpoint", "server")
+            dev.set_param("dir", self._image_params['nbd_server_tls_creds'])
+            ret = self._vm.devices.simple_hotplug(dev, self._vm.monitor)
+            if not ret[1]:
+                raise exceptions.TestFail("Failed to hotplug device '%s': %s."
+                                          % (dev, ret[0]))
+
+    def start_nbd_server(self):
+        server = {
+            'type': 'unix',
+            'path': self._image_params['nbd_unix_socket']
+        } if self._image_params.get('nbd_unix_socket') else {
+            'type': 'inet',
+            'host': '0.0.0.0',
+            'port': self._image_params.get('nbd_port', '10809')
+        }
+
+        logging.info("Start internal nbd server")
+        return self._vm.monitor.nbd_server_start(server, self._tls_creds_id)
+
+    def add_nbd_image(self, node_name=None):
+        """
+        :param node_name: block node name, the node might be hotplugged
+                          by other utils, or the node has already been
+                          present in VM.
+        """
+        if node_name:
+            self._node_name = node_name
+
+        logging.info("Add image node to nbd server")
+        return self._vm.monitor.nbd_server_add(
+            self._node_name,
+            self._image_params.get('nbd_export_name'),
+            self._image_params.get('nbd_export_writable'),
+            self._image_params.get('nbd_export_bitmap'))
+
+    def remove_nbd_image(self):
+        logging.info("Remove image from nbd server")
+        return self._vm.monitor.nbd_server_remove(
+            self._node_name, self._image_params.get('nbd_remove_mode')
+        )
+
+    def export_image(self):
+        self.start_nbd_server()
+        self.add_nbd_image()
+
+    def stop_export(self):
+        logging.info("Stop nbd server")
+        return self._vm.monitor.nbd_server_stop()
```

Comment:
I think `def stop_nbd_server(self):` is more readable, since` logging.info("Stop nbd server") `and `monitor.nbd_server_stop()`, it could be corresponding with `start_nbd_server`

Reply:
added a new function stop_nbd_server, and leave stop_export here which calls stop_nbd_server, for stop_export do the following:
1. unregister all devices
2. stop QEMU's embedded NBD server

these two steps are performed with one qmp command nbd-server-stop, encapsulated in vm.monitor.nbd_server_stop()